### PR TITLE
Update product status after calling save() on a new product

### LINKED ISSUE #48290: WC_Product::get_status() not correct after calling save() on new product
**Issue Description:**

When a new product is created without setting a status, the product is published. However, this is not reflected in the status returned by `WC_Product::get_status()`. For example:

```php
$product = new WC_Product();
$product->save();
$status = $product->get_status();
// $status is `false`, when it should be `"publish"`
```

This bug makes using the product object after the save unreliable. For example, if the `woocommerce_new_product` action is used, the product passed to that action will not have the correct status, which can cause incorrect behavior. I found this while working on:

- https://github.com/woocommerce/woocommerce/pull/48194

### PR DESCRIPTION
**Pull request by mattsherman**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR updates the implementation of the `WC_Product_Data_Store_CPT::create()` method to update the product's status after the product has been saved, so that the `WC_Product::get_status()` method returns the correct status for a new product created without explicitly setting the status:

```php
$product = new WC_Product();
$product->save();
$status = $product->get_status();
// $status should be `"publish"`
```

Closes #48290.

### How to test the changes in this Pull Request:

1. Run the unit tests (or just look at the CI results on this PR) and verify they all pass (and make sense):

```
pnpm test:unit:env --filter 'WC_Tests_Product_Data_Store'
```

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php [hunk 1]
```diff
@@ -138,6 +138,12 @@ public function create( &$product ) {
 		if ( $id && ! is_wp_error( $id ) ) {
 			$product->set_id( $id );
 
+			// get the post object so that we can set the status
+			// to the correct value; it is possible that the status was
+			// changed by the woocommerce_new_product_data filter above.
+			$post_object = get_post( $product->get_id() );
+			$product->set_status( $post_object->post_status );
+
```
Comment (mattsherman) [inline plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:146]: I was tempted to just call `$this->read( $product );` after the call to `$product->apply_changes()` below, to ensure that everything about the product reflects what was stored in the DB. But, that is a much bigger change in my opinion, has more room for subtle backwards-compatibility issues, and might have more of a performance impact. I still think that change should be made, but that is not immediately needed (while this is, for #48194).

**Comment by mattsherman** [comment]
@woocommerce/proton Since this PR makes a small, yet potentially impactful, change to the `WC_Product` CRUD implementation, I'd love another set of eyes on this from outside Mothra to make sure this implementation is sound. Thanks!

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 1 above)
Comment (barryhughes) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:146]: Gathering some thoughts on this, not immediately sure what is best. An existing pattern used elsewhere, though, basically takes this form:

```php
$product = new WC_Product;
$product->save();
$product = wc_get_product( $product->get_id() );
```

I'm not suggesting that is ideal, or what most would expect to have to do, though.

**Review by barryhughes — COMMENTED** [review]
Following up:

> I was tempted to just call `$this->read( $product );` after the call to `$product->apply_changes()` below, to ensure that everything about the product reflects what was stored in the DB.

The consensus seems to be we should do exactly this. Otherwise, things are still inconsistent (just a slightly different set of inconsistencies than at present).

However, it feels like we could probably also move ahead with this in its current form and log a follow-up referencing this change, to do the wider bit of work at some later point (and support it with additional testing, etc).

**Comment by mattsherman** [comment]
@barryhughes Thanks!

> Following up:
> 
> > I was tempted to just call `$this->read( $product );` after the call to `$product->apply_changes()` below, to ensure that everything about the product reflects what was stored in the DB.
> 
> The consensus seems to be we should do exactly this. Otherwise, things are still inconsistent (just a slightly different set of inconsistencies than at present).

Agreed.

> However, it feels like we could probably also move ahead with this in its current form and log a follow-up referencing this change, to do the wider bit of work at some later point (and support it with additional testing, etc).

Let's do that. It seems less risky.

**Review by barryhughes — APPROVED** [review]
Nothing else to add, looks good to me. 

✅ Product updates and creation continues to work as before (both when interacting with the catalog from the admin UI, and when interacting via the REST API).

🧪 Thank you for adding the extra assertions!

Approving but not merging, in case you wanted one of the other reviewers to take a peek (if that's not needed, please do go ahead and merge) 👍🏼

**Comment by mattsherman** [comment]
> log a follow-up referencing this change, to do the wider bit of work at some later point (and support it with additional testing, etc).

Logged:

- https://github.com/woocommerce/woocommerce/issues/48547

**Pull request merged**